packages/docker/src/plugins/plugin.ts

```diff
@@ -0,0 +1,181 @@
+import {
+  type CreateNodesV2,
+  type ProjectConfiguration,
+  type TargetConfiguration,
+  createNodesFromFiles,
+  readJsonFile,
+  writeJsonFile,
+  CreateNodesContext,
+} from '@nx/devkit';
+import { calculateHashesForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
+import { hashObject } from 'nx/src/hasher/file-hasher';
+import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';
+import { existsSync } from 'fs';
+import { dirname, join } from 'path';
+import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
+
+export interface DockerPluginOptions {
+  buildTarget?: string;
+  runTarget?: string;
+}
+
+interface NormalizedDockerPluginOptions {
+  buildTarget: string;
+  runTarget: string;
+}
+
+type DockerTargets = Pick<ProjectConfiguration, 'targets' | 'metadata'>;
+
+function readTargetsCache(cachePath: string): Record<string, DockerTargets> {
+  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
+}
+
+function writeTargetsCache(
+  cachePath: string,
+  results?: Record<string, DockerTargets>
+) {
+  writeJsonFile(cachePath, results ?? {});
+}
+
+const dockerfileGlob = '**/Dockerfile';
+
+export const createNodesV2: CreateNodesV2<DockerPluginOptions> = [
+  dockerfileGlob,
+  async (configFilePaths, options, context) => {
+    const optionsHash = hashObject(options);
+    const cachePath = join(
+      workspaceDataDirectory,
+      `docker-${optionsHash}.hash`
+    );
+    const targetsCache = readTargetsCache(cachePath);
+    const projectRoots = configFilePaths.map((c) => dirname(c));
+    const normalizedOptions = normalizePluginOptions(options);
+    const hashes = await calculateHashesForCreateNodes(
+      projectRoots,
```

Comment:
nit for later, we don't need to hash the entire project directory, only the dockerfile. See if there's a way we can do that.